File: test/test_F2E_UR.js
// Port of test/test_F2E_UR.py — assert that every scramble in F2E_UR.txt
// produces the F2E shape: 14 targets, no flips, and a final 1-target cycle
// with orientation=1, parity=1.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import { weakswapEdgesUR } from './test_tracer.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function main() {
    const scramblesFile = path.join(__dirname, 'F2E_UR.txt');
    const scrambles = readFileSync(scramblesFile, 'utf8')
        .split('\n')
        .map((l) => l.trim())
        .filter((l) => l.length > 0);

    for (let i = 0; i < scrambles.length; i++) {
        const scram = scrambles[i];
        const result = weakswapEdgesUR(scram);
        const targetsCount = result.targets.length;
        const flipsCount = result.flips.length;
        const remainingCycles = result.remaining_cycles_sorted;

        const fail = (msg) => {
            console.log(`FAIL at scramble ${i + 1}: ${scram}`);
            console.log(`Error: ${msg}`);
            console.log(`Targets (${targetsCount}): ${JSON.stringify(result.targets)}`);
            console.log(`Flips (${flipsCount}): ${JSON.stringify(result.flips)}`);
            console.log(`Remaining cycles sorted: ${JSON.stringify(remainingCycles)}`);
            process.exit(1);
        };

        if (targetsCount !== 14) fail(`Expected 14 targets, got ${targetsCount}`);
        if (flipsCount !== 0) fail(`Expected 0 flips, got ${flipsCount}`);

        if (remainingCycles.length > 0) {
            const shortest = remainingCycles[remainingCycles.length - 1];
            const shortestLen = shortest.targets.length;
            const shortestOri = shortest.orientation;
            const shortestParity = shortest.parity;
            if (shortestLen !== 1) fail(`Expected shortest cycle to have 1 target, got ${shortestLen}`);
            if (shortestOri !== 1) fail(`Expected shortest cycle orientation 1, got ${shortestOri}`);
            if (shortestParity !== 1) fail(`Expected shortest cycle parity 1, got ${shortestParity}`);
        }
    }

    console.log(
        `PASS: All ${scrambles.length} scrambles have 14 targets, 0 flips, and shortest cycle has 1 target with ori=1 parity=1`
    );
}

main();

File: test/test_F2E_UR.py
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from test_tracer import weakswap_edges_UR

def main():
    scrambles_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "F2E_UR.txt")

    with open(scrambles_file, "r") as f:
        scrambles = [line.strip() for line in f if line.strip()]

    for i, scram in enumerate(scrambles):
        result = weakswap_edges_UR(scram)
        targets_count = len(result["targets"])
        flips_count = len(result["flips"])
        remaining_cycles = result["remaining_cycles_sorted"]

        try:
            assert targets_count == 14, f"Expected 14 targets, got {targets_count}"
            assert flips_count == 0, f"Expected 0 flips, got {flips_count}"

            # Check shortest cycle (last in sorted list) has 1 target, orientation 1, parity 1
            if remaining_cycles:
                shortest_cycle = remaining_cycles[-1]
                shortest_targets_len = len(shortest_cycle["targets"])
                shortest_ori = shortest_cycle["orientation"]
                shortest_parity = shortest_cycle["parity"]
                assert shortest_targets_len == 1, f"Expected shortest cycle to have 1 target, got {shortest_targets_len}"
                assert shortest_ori == 1, f"Expected shortest cycle orientation 1, got {shortest_ori}"
                assert shortest_parity == 1, f"Expected shortest cycle parity 1, got {shortest_parity}"
        except AssertionError as e:
            print(f"FAIL at scramble {i + 1}: {scram}")
            print(f"Error: {e}")
            print(f"Targets ({targets_count}): {result['targets']}")
            print(f"Flips ({flips_count}): {result['flips']}")
            print(f"Remaining cycles sorted: {remaining_cycles}")
            sys.exit(1)

    print(f"PASS: All {len(scrambles)} scrambles have 14 targets, 0 flips, and shortest cycle has 1 target with ori=1 parity=1")

if __name__ == "__main__":
    main()